src/worker/activities/activity_heartbeat_manager.rs

```diff
@@ -105,143 +99,173 @@ impl ActivityHeartbeatManager {
 }
 
 #[derive(Debug)]
-struct HbStreamState {
-    last_sent: HashMap<TaskToken, ActivityHbState>,
-    incoming_hbs: UnboundedReceiver<HBAction>,
+struct ActivityHbState {
+    last_recorded_details: Option<Vec<common::Payload>>,
+    shutdown_tx: UnboundedSender<()>,
 }
 
-impl HbStreamState {
-    fn new(incoming_hbs: UnboundedReceiver<HBAction>) -> Self {
-        Self {
-            last_sent: Default::default(),
-            incoming_hbs,
+async fn send_heartbeat(
+    sg: Arc<impl ServerGatewayApis + Send + Sync + 'static + ?Sized>,
+    cancels_tx: UnboundedSender<PendingActivityCancel>,
+    task_token: TaskToken,
+    details: Vec<common::Payload>,
+) {
```

Comment:
Why make such dramatic changes to the entire structure here? The existing implementation had the nice property that there aren't bunches of `Arc`s and locks.

I think the streaming approach could pretty easily be kept while still accomplishing the goal